nav2_costmap_2d/plugins/costmap_filters/speed_filter.cpp

```diff
@@ -0,0 +1,372 @@
+  const geometry_msgs::msg::Pose2D & pose,
+  geometry_msgs::msg::Pose2D & mask_pose) const
+{
+  rclcpp_lifecycle::LifecycleNode::SharedPtr node = node_.lock();
+  if (!node) {
+    throw std::runtime_error{"Failed to lock node"};
+  }
+
+  if (mask_frame_ != global_frame_) {
+    // Filter mask and current layer are in different frames:
+    // Prepare transform from current layer frame (global_frame_) to mask frame
+    geometry_msgs::msg::TransformStamped transform;
+    tf2::Transform tf2_transform;
+    tf2_transform.setIdentity();  // initialize by identical transform
+    try {
+      transform = tf_->lookupTransform(
+        mask_frame_, global_frame_, tf2::TimePointZero,
+        transform_tolerance_);
+    } catch (tf2::TransformException & ex) {
+      RCLCPP_ERROR(
+        node->get_logger(),
+        "SpeedFilter: failed to get costmap frame (%s) "
+        "transformation to mask frame (%s) with error: %s",
+        global_frame_.c_str(), mask_frame_.c_str(), ex.what());
+      return false;
+    }
+    tf2::fromMsg(transform.transform, tf2_transform);
+
+    // Transform (pose.x, pose.y) point from global_frame_
+    // to mask_pose point in mask_frame_
+    tf2::Vector3 point(pose.x, pose.y, 0);
+    point = tf2_transform * point;
+    mask_pose.x = point.x();
+    mask_pose.y = point.y();
+  } else {
+    // Filter mask and current layer are in the same frame:
+    // Just use pose coordinates
+    mask_pose = pose;
+  }
+
+  return true;
+}
+
+bool SpeedFilter::worldToMask(double wx, double wy, unsigned int & mx, unsigned int & my) const
+{
+  double origin_x = filter_mask_->info.origin.position.x;
+  double origin_y = filter_mask_->info.origin.position.y;
+  double resolution = filter_mask_->info.resolution;
+  unsigned int size_x = filter_mask_->info.width;
+  unsigned int size_y = filter_mask_->info.height;
+
+  if (wx < origin_x || wy < origin_y) {
+    return false;
+  }
+
+  mx = std::round((wx - origin_x) / resolution);
+  my = std::round((wy - origin_y) / resolution);
+  if (mx >= size_x || my >= size_y) {
+    return false;
+  }
+
+  return true;
+}
+
+void SpeedFilter::process(
+  nav2_costmap_2d::Costmap2D & /*master_grid*/,
+  int /*min_i*/, int /*min_j*/, int /*max_i*/, int /*max_j*/,
+  const geometry_msgs::msg::Pose2D & pose)
+{
+  std::lock_guard<CostmapFilter::mutex_t> guard(*getMutex());
+
+  rclcpp_lifecycle::LifecycleNode::SharedPtr node = node_.lock();
+  if (!node) {
+    throw std::runtime_error{"Failed to lock node"};
+  }
+
+  if (!filter_mask_) {
+    // Show warning message every 2 seconds to not litter an output
+    RCLCPP_WARN_THROTTLE(
+      node->get_logger(), *(node->get_clock()), 2000,
+      "Filter mask was not received");
+    return;
+  }
+
+  geometry_msgs::msg::Pose2D mask_pose;  // robot coordinates in mask frame
+
+  // Transforming robot pose from current layer frame to mask frame
+  if (!transformPose(pose, mask_pose)) {
+    // Robot pose is unknown. Returning.
```

Comment:
This comment I think is a little TMI, I think the `if not transform pose -> return` is sufficiently clear with the comment 2 lines up.

Reply:
Removed in https://github.com/ros-planning/navigation2/commit/3465c90415afd11418f98fc01bbc9386afa853ea#diff-d344e376ecbc8e5ecd5045731e13855c6d2a9a955cdaee9b31dddbbee7681466L276 and https://github.com/ros-planning/navigation2/commit/3465c90415afd11418f98fc01bbc9386afa853ea#diff-d344e376ecbc8e5ecd5045731e13855c6d2a9a955cdaee9b31dddbbee7681466L283